source/Armament.cpp

```diff
@@ -207,7 +207,7 @@ int Armament::TurretCount() const
 set<const Outfit *> Armament::RestockableAmmo() const
 {
 	auto restockable = set<const Outfit *>{};
-	for(auto &&hardpoint : Get())
+	for(const Hardpoint &hardpoint : hardpoints)
```

Comment:
Going from `Get()` to `hardpoints` is all fine, but I will admit that I've never been able to figure out the difference between iterating with an rvalue reference (&&) and iterating with just a normal reference (&). We have a fair few uses of iterating over rvalue references in the codebase, which I believe have mainly been introduced by tehhowch.

Reply:
rvalue means you can change the object you are pointing to, which is not necessary here, it was necessary because get returned a reference already but with this there is no use to it anymore

Reply:
> rvalue means you can change the object you are pointing to

So then what would the distinction be between `auto &&it : vector` and `auto &it : vector`? Both can be changed as neither are const.

Reply:
an rvalue is not about changing the variable but changing the variable you are pointing to so if you have been pointing to memory address 0x00 before you can change it to 0x01 for example which would be impossible with a standard reference where the address is constant, but it is totally unnecessary here

Reply:
I don't think the line here is actually producing an rvalue reference anyway.
`Get()` returns a `const vector<Hardpoint> &`
VSCode intellisense suggests to me that `hardpoint` here is a `const Hardpoint &`.
If I remove the `auto` and specifically do:
`const Hardpoint &&hardpoint : Get()` I get an error. An rvalue reference cannot be bound to an lvalue.

Reply:
> an rvalue is not about changing the variable but changing the variable you are pointing to so if you have been pointing to memory address 0x00 before you can change it to 0x01 for example which would be impossible with a standard reference where the address is constant, but it is totally unnecessary here

Still not following on what the difference is for this case. I have an understanding of how rvalue references are used generally, but I'm still not sure what the difference is for iterating over a container. I can recompile this code with `const Hardpoint &hardpoint : Get()`, as well as with `auto &&hardpoint : hardpoints`, so how we're getting the container that we're iterating over seems inconsequential to whether using an rvalue reference or a const reference works. You say "it was necessary" but now "it's totally unnecessary here" due to swapping `Get()` with `hardpoints`, but it seems to always have been unnecessary.

(Not saying this line needs to change for this PR. I've just been digging into this rvalue reference stuff for one of my PRs and I'm taking the opportunity to ask questions.)

Reply:
from the cpp reference:

> If the placeholder type specifier is auto or type-constraint auto(since C++20), the variable type is deduced from the initializer using the rules for [template argument deduction](https://en.cppreference.com/w/cpp/language/template_argument_deduction) from a function call (see [template argument deduction — other contexts](https://en.cppreference.com/w/cpp/language/template_argument_deduction#Other_contexts) for details).
> For example, given const auto& i = expr;, the type of i is exactly the type of the argument u in an imaginary template template<class U> void f(const U& u) if the function call f(expr) was compiled. Therefore, **auto&& may be deduced either as an lvalue reference or rvalue reference according to the initializer**, which is used in range-based for loop.

it seems that this is no rvalue at all but just an easy way to say it could be lvalue or rvalue if you are too lazy to check what it is